pkg/volume/rbd/attacher.go

```diff
@@ -213,13 +234,35 @@ func (detacher *rbdDetacher) UnmountDevice(deviceMountPath string) error {
 		return err
 	}
 	klog.V(3).Infof("rbd: successfully umount device mountpath %s", deviceMountPath)
+	klog.V(4).Infof("rbd: detaching device %s", mp.Device)
 
-	klog.V(4).Infof("rbd: detaching device %s", devicePath)
-	err = detacher.manager.DetachDisk(detacher.plugin, deviceMountPath, devicePath)
+	// add some retries for detach operation
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Steps:    5,
+	}
+	wait.ExponentialBackoff(backoff, func() (bool, error) {
+		err = detacher.manager.DetachDisk(detacher.plugin, deviceMountPath, mp.Device)
+		if err == nil {
+			return true, nil
+		}
+		klog.Errorf("rbd: detach device failed: %s, %v, will try again.", mp.Device, err)
+		return false, nil
+	})
 	if err != nil {
+		klog.Errorf("rbd: detach device failed: %s, %s, try to mount device back: %v", mp.Device, err, *mp)
+		// Try to mount back the device, otherwise the next UnmountDevice call will get an empty DevicePath,
+		// then the device will get no chance to be unmounted.
+		mErr := detacher.mounter.Mount(mp.Device, mp.Path, mp.Type, mp.Opts)
```

Comment:
Just as @jsafrane said, I don't like to mount it back. This is not the right way to fix it, though this could help mitigate the problem. 

I think we should dig deeper into `(util *RBDUtil) DetachDisk` to see why the command gets failed. Actually this part has nothing to do with kubelet. It is purely RBD part. More precisely, we should care more about what commands and arguments we issued to `rbd`.